VariantShredding.md

```diff
@@ -25,290 +25,316 @@
+|----------|---------------|-------------------------------------------------------------|
+| null     | null          | The value is missing; only valid for shredded object fields |
+| non-null | null          | The value is present and may be any type, including null    |
+| null     | non-null      | The value is present and is the shredded type               |
+| non-null | non-null      | The value is present and is a partially shredded object     |
 
-Similarly the elements of an `array` must be a group containing one or more of `object`, `array`, `typed_value` or `variant_value`.
+An object is _partially shredded_ when the `value` is an object and the `typed_value` is a shredded object.
 
-Each leaf in the schema can store an arbitrary Variant value.
-It contains an `variant_value` binary field and a `typed_value` field.
-If non-null, `variant_value` represents the value stored as a Variant binary.
-The `typed_value` field may be any type that has a corresponding Variant type.
-For each value in the data, at most one of the `typed_value` and `variant_value` may be non-null.
-A writer may omit either field, which is equivalent to all rows being null.
+If both fields are non-null and either is not an object, the value is invalid. Readers must either fail or return the `typed_value`.
 
-Dictionary IDs in a `variant_value` field refer to entries in the top-level `metadata` field.
+If a Variant is missing in a context where a value is required, readers must either fail or return a Variant null: basic type 0 (primitive) and physical type 0 (null).
+For example, if a Variant is required (like `measurement` above) and both `value` and `typed_value` are null, the returned `value` must be `00` (Variant null).
 
-For an `object`, a null field means that the field does not exist in the reconstructed Variant object.
-All elements of an `array` must be non-null, since array elements cannote be missing.
+### Shredded Value Types
 
-| typed_value | variant_value | Meaning |
-|-------------|----------------|---------|
-| null | null | Field is Variant Null (not missing) in the reconstructed Variant. |
-| null | non-null | Field may be any type in the reconstructed Variant. |
-| non-null | null | Field has this column’s type in the reconstructed Variant. |
-| non-null | non-null | Invalid |
+Shredded values must use the following Parquet types:
 
-The `typed_value` may be absent from the Parquet schema for any field, which is equivalent to its value being always null (in which case the shredded field is always stored as a Variant binary).
-By the same token, `variant_value` may be absent, which is equivalent to their value being always null (in which case the field will always have the value Null or have the type of the `typed_value` column).
+| Variant Type                | Equivalent Parquet Type      |
+|-----------------------------|------------------------------|
+| boolean                     | BOOLEAN                      |
```

Comment:
I think this goes back to earlier debates about allowed lossiness at the storage layer but it seems we might lose a lot of benefits to shredding if a single variant has multiple exact integer representations (e.g. mix of int32 and int64)

Reply:
I think the state of this is that the storage will not do any conversions at all. However, the engine itself is allowed to "normalize" variants to optimize. In this case, engines will probably normalize within "exact numerics" in order to make shredding more effective.

Reply:
As I digest the types, I feel like the concept of exact numeric is purely an engine concern and is an entirely new concept for parquet that isnt really explicitly modelled anyplace in the spec.  Given this I think my preferences would be:

1.  Remove the concept of this type of equality  from the parquet spec and leave it in the engine.
2. Make conversion to JSON an appendix as a recommendation 
3. Make sure we design the variant in parquet to optimized with shredding independent from  the engine (I think this means adding a type identifier discriminant field for the shredded column) also as I stated in another comment, it would be good to ensure the bit order for values makes it so that we can make the widest use of stats (e.g. type info is in the most significant bits) for the non-shredded value (iiuc this might require a second version for the non-shredded value, depending on what the current bit order means) 

Thoughts?

Reply:
The intent is to leave normalization and conversion to engines, but I don't think that we need to remove the Variant logical types in order to do that.

I also think that we need to have a reliable JSON representation or else we will get different behavior and conventions across engines. We want people to be able to rely on what the JSON means, even if it is necessarily lossy in some cases.

Reply:
> The intent is to leave normalization and conversion to engines, but I don't think that we need to remove the Variant logical types in order to do that.

I think this might just be a philosophical difference of opinion.  But I believe Parquet implementations should have the flexibility to appropriately optimize across physical type.

 I think we should either:
1.  Not have the notion of logical type correspondence as part of variant in parquet.  Under this assumption we should provide enough modeling that parquet can do a reasonable job shredding and reproducing the exact physical types after shredding.  Parquet is used in a lot of contexts and I think in many cases the engines might not be very sophisticate here.
2.  Not require parquet to reproduce the exact physical types when reading back values (i.e. allow it to normalization).

Both approaches have there strengths, but what I would like to avoid is a tight coupling between the engine and parquet libraries to achieve efficient storage of variants.

Reply:
The original motivation for introducing the "logical type" for the Variant binary encoding was to separate the "logical" data value, from the "physical" storage of the value, so that optimizations may be performed without altering the "logical" value. The value is important, not the storage type. For example, it shouldn't matter if the string `"hello"` is stored as a short-string or a string, and it shouldn't matter if the integer `1` is stored as an `int8` or `int16`. Within the same "logical type" the equivalent value can be stored in different ways.

We haven't fully defined the equality semantics yet, but the goal for the logical types is that if we had 2 Variant values:
- `"hello"` (as a short-string)
- `"hello"` (as a string)
those Variant values are considered equivalent, even though they are physically encoded differently.

Reply:
@gene-db This makes sense.  I think we need to decide at the parquet level to either:
1.  Use this definition of logical type and make it valid for parquet to do normalization (e.g. physical types put in are not necessarily physical types you get out).
2. Make it possible for parquet to do lossless normalization (e.g. store values in the widest type that makes sense, and then also store metadata to retrieve types).

Reply:
> Use this definition of logical type and make it valid for parquet to do normalization (e.g. physical types put in are not necessarily physical types you get out).

My opinion is that parquet the format should not "do" normalization (though it would be fine if the writer implementations did)

> I think the state of this is that the storage will not do any conversions at all. However, the engine itself is allowed to "normalize" variants to optimize. 

I believe this writeup from @gene-db describes what I would expect from a formrat. In summary:
1. The spec itself doesn't require / permit any normalization of values
2. The implementations of writers (e.g. parquet-rs or java-rs) could have options to automatically coerce types 

My rationale is that one normalization scheme is unlikely to work for all usecases, so mandating something in the spec will unnecessarily constrain users

@etseidl, @dsgibbons and others have been heading this way in parquet-rs recently (for example https://github.com/apache/arrow-rs/pull/6828) when normalizing data / schema might be preferred for wider compatibility, but users can still choose to use the raw data and maintain lossless roundtrip)